lib/db/schema/patch-065-066.sql

```diff
@@ -0,0 +1,83 @@
+SET NAMES utf8mb4 COLLATE utf8mb4_bin;
+
+-- Add an index on sessionTokens::createdAt to fix pruning filesort.
+CREATE INDEX `sessionTokens_createdAt`
+ON `sessionTokens` (`createdAt`)
+ALGORITHM=INPLACE
+LOCK=NONE;
+
+-- Used to prevent session token pruning from doing a full table scan
+-- as it proceeds further and further through the (very long) backlog
+-- of pruning candidates.
+INSERT INTO dbMetadata (name, value)
+VALUES ('sessionTokensPrunedUntil', 0);
+
+-- Update the prune stored procedure with assorted jiggery-pokery to
+-- make session token pruning faster.
+CREATE PROCEDURE `prune_5` (IN `olderThan` BIGINT UNSIGNED)
+BEGIN
+  DECLARE EXIT HANDLER FOR SQLEXCEPTION
+  BEGIN
+    ROLLBACK;
+    RESIGNAL;
+  END;
+
+  SELECT @lockAcquired := GET_LOCK('fxa-auth-server.prune-lock', 3);
+
+  IF @lockAcquired THEN
+    DELETE FROM accountResetTokens WHERE createdAt < olderThan ORDER BY createdAt LIMIT 10000;
+    DELETE FROM passwordForgotTokens WHERE createdAt < olderThan ORDER BY createdAt LIMIT 10000;
+    DELETE FROM passwordChangeTokens WHERE createdAt < olderThan ORDER BY createdAt LIMIT 10000;
+    DELETE FROM unblockCodes WHERE createdAt < olderThan ORDER BY createdAt LIMIT 10000;
+    DELETE FROM signinCodes WHERE createdAt < olderThan ORDER BY createdAt LIMIT 10000;
+
+    START TRANSACTION;
+
+    -- Jiggery-pokery #1: Find out how far we got on previous iterations.
+    SELECT @newerThan := value FROM dbMetadata WHERE name = 'sessionTokensPrunedUntil';
+
+    -- Jiggery-pokery #2: Store the pruned session tokens in a temporary table
+    --                    ahead of time, so we can refer to createdAt post-facto.
+    CREATE TEMPORARY TABLE prunedSessionTokens (
+      tokenId BINARY(32) PRIMARY KEY,
+      createdAt BIGINT UNSIGNED NOT NULL,
+      INDEX prunedSessionTokens_createdAt (createdAt)
+    ) AS
+      SELECT tokenId, createdAt FROM sessionTokens
+      WHERE createdAt >= @newerThan AND createdAt < olderThan
+      AND NOT EXISTS (
+        SELECT sessionTokenId FROM devices
+        WHERE sessionTokenId = sessionTokens.tokenId
+      )
+      ORDER BY createdAt
+      LIMIT 10000;
```

Comment:
With the caveat that I've no idea whether it'll work, I wonder if we can avoid the temp table here by doing a select to find the time range, then deleting from that time range, then storing the done range in the table.  Something like:

* `pruneUpTo = SELECT MAX(createdAt) FROM (SELECT createdAt FROM sessionTokens WHERE <no device record> ORDER BY `createdAt` LIMIT 10000`
* `DELETE FROM sessionTokens WHERE createdAt > $sessionTokensPrunedUntil AND createdAt < $pruneUpTo`
* Update $sessionTokensPrunedUntil with the value from $pruneUpTo

Reply:
IIUC this would mean doing the lookup in `devices` twice though, right? We couldn't do the `DELETE` as you've typed it because that might get rid of `sessionTokens` with device records? Or have I misunderstood.

Anyway, I can have a play and try it out. I have a local db with 10 millions session tokens. Nothing like the production count but it should still be useful for comparing how different possibilities compare against each other.

Reply:
Fair play, ditching the temporary table appears to make the query 2 to 3 times faster in my tests.

I've pushed it as a separate commit, 4a1f258, to make it easier to see what I changed. Hopefully it's what you had in mind.

I assumed doing the lookup in `devices` twice would make it much slower, is MySQL caching that because it's all in the same transaction? Or is it just less punitive than the overhead of writing a temporary table?